File: 08-Graph/jooyoung/7569_토마토.py
''''
보관 후 하루가 지나면,
익은 토마토들의 인접한 곳에 있는 익지 않은 토마토들은
익은 토마토의 영향을 받아 익게 된다.
'''

M, N, H = map(int, input().split())  # 세로, 가로, 상자 수
box = [[list(map(int, input().split())) for _ in range(N)] for _ in range(H)]
visited = [[[False] * M for _ in range(N)] for _ in range(H)]
q = []
answer = 0


def bfs():
    while q:
        x, y, z = q.pop(0) # 익은 토마토의 위치
        for o in range(6):  # 6개의 방향을 확인해준다
            nx = x + dx[o]
            ny = y + dy[o]
            nz = z + dz[o]

            if nx < 0 or nx >= H or ny < 0 or ny >= N or nz < 0 or nz >= M:
                continue

            if box[nx][ny][nz] == 0 and not visited[nx][ny][nz]:
                q.append((nx, ny, nz))  # 익지 않은 토마토
                box[nx][ny][nz] = box[x][y][z] + 1  # 익은 토마토로 변경
                visited[nx][ny][nz] = True  # 방문 표시


dx = [-1, 1, 0, 0, 0, 0]
dy = [0, 0, 1, -1, 0, 0]
dz = [0, 0, 0, 0, 1, -1]

for i in range(H):
    for j in range(N):
        for k in range(M):
            if box[i][j][k] == 1 and not visited[i][j][k] :
                q.append((i, j, k))  # 배열을 도는 동안 익은 토마토가 있고 방문 한적이 없다면 큐에 삽입
                visited[i][j][k] = True

bfs()

for i in box:
    for j in i:
        for k in j:
            if k == 0:
                print(-1)
                exit(0)
        answer = max(answer, max(j))

print(answer - 1)

File: 08-Graph/jaehoon/7569_토마토.cpp
#include<iostream>
#include<cstring>
#include<vector>
#include<queue>
#include <utility>


using namespace std;

int dz[6] = { 0,0,0,0,-1,1 };
int dx[6] = { 0,0,1,-1,0,0 };
int dy[6] = { 1,-1,0,0,0,0 };

int M, N,H;

int visited[101][101][101] = { 0 }; // bfs활용시 날짜를 계산하는 배열 (dist로 할껄 그랬습니다..)
int tomato[101][101][101]; //토마토 입력 받는 배열
int result = 0;

queue<pair<pair<int, int>, int>> q;

void bfs() { //q가 빌때까지 순회하는 함수

	while (!q.empty()) {
		pair<pair<int, int>, int> cur = q.front(); 
		q.pop();

		for (int i = 0; i < 6; i++) {
			//현재 좌표에 x,y,z를 이동한 각각의 좌표를 만드는 코드
			int nz = cur.first.first + dz[i];
			int nx = cur.first.second + dx[i];
			int ny = cur.second + dy[i];

			//주변에 익지 않은 토마토가 있으면 그 전 토마토가 익은 날짜에 +1해줌 그리고 그 토마토 위치 다시 push
			if (nx >= 0 && nx < N && ny >= 0 && ny < M && nz >= 0 && nz < H) {
				if (visited[nz][nx][ny]==-1) {
					visited[nz][nx][ny] = visited[cur.first.first][cur.first.second][cur.second] + 1;
					q.push({ {nz,nx},ny });
				}
			}
		}
	}

}

int main() {

	cin >> M>>N>>H;

	//토마토 넣는 로직 
	for (int k = 0; k < H; k++) {
		for (int i = 0; i < N; i++) {
			for (int j = 0; j < M; j++) {
				cin >> tomato[k][i][j];
				if (tomato[k][i][j]==1) { //익은 토마토가 존재하면 q에 push
					q.push({ { k,i }, j });
				}
				if (tomato[k][i][j] == 0) { //안익은 토마토 존재하면 visited -1 (가중치)
					visited[k][i][j] = -1;
				}
			}
		}
	}
	
	bfs();


	for (int k = 0; k < H; k++) {
		for (int i = 0; i < N; i++) {
			for (int j = 0; j < M; j++) {
				if (visited[k][i][j] == -1) { //하나라도 안익었으면 -1 출력
					cout << -1;
					return 0;
				}
				result = max(result, visited[k][i][j]);
			}
		}
	}

	cout << result;



}